Download voices from NVDA version (experimental)

rmcpantoja opened the pull request:
This PR

- Adds a voice downloader for NVDA addon. The interphase is similar to Android and works in a package directory.

This is still WIP, but comments are welcome.

zstanecic commented:
@rmcpantoja 
This is even not building on windows...
scons: *** [build\windows\packages\nvda\RHVoice-1.16.3.nvda-addon] Source `src\nvda-addon\globalPlugin\RHVoice_globalPlugin.py' not found, needed by target `build\windows\packages\nvda\RHVoice-1.16.3.nvda-addon'.
scons: building terminated because of errors

zstanecic commented:
Not really. still not buildable.
AttributeError: <class 'SCons.Node.FS.Dir'> object has no attribute 'Glob':                                             
  File "C:\RHVoice\SConstruct", line 387:                                                                               
    build_for_windows(base_env,vars)                                                                                    
  File "C:\RHVoice\SConstruct", line 375:                                                                               
    SConscript(os.path.join("src","nvda-addon","SConscript"),                                                           
  File "C:\Program Files (x86)\Python311-32\Lib\site-packages\SCons\Script\SConscript.py", line 676:                    
    return method(*args, **kw)                                                                                          
  File "C:\Program Files (x86)\Python311-32\Lib\site-packages\SCons\Script\SConscript.py", line 612:                    
    return _SConscript(self.fs, *files, **subst_kw)                                                                     
  File "C:\Program Files (x86)\Python311-32\Lib\site-packages\SCons\Script\SConscript.py", line 279:                    
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)                                               
  File "C:\RHVoice\src\nvda-addon\SConscript", line 34:                                                                 
    addon_files = curdir.Dir("globalPlugin").Dir("RHVoice").Glob("**/*")                                                
  File "C:\Program Files (x86)\Python311-32\Lib\site-packages\SCons\Node\FS.py", line 672:                              
    raise AttributeError("%r object has no attribute %r" %                                                              
P

rmcpantoja commented:
Hi @zstanecic. 
Thanks for the build testing! I just fixed it again. You should have the generated addon without errors anymore.
Please test again and send your feedback.
Cheers.

zstanecic commented:
Okay: you have a very big logical error.
When you install just the speech synthesizer driver, you cannot use the downloader. It tells you, that you should set the synthesizer to RHVoice. How do you want to achieve this, if you have no voices installed?

rmcpantoja commented:
> Okay: you have a very big logical error. When you install just the speech synthesizer driver, you cannot use the downloader. It tells you, that you should set the synthesizer to RHVoice. How do you want to achieve this, if you have no voices installed?

Hi @zstanecic. 
We need to use the NVDA's synthDriverHandler under RHVoice module to capture how many voices are installed, and what are all of these. This can be used for update capabilities in the future, depending if this is of sufficient interest.
But, for the first run / if the user doesn't have voices,

- a) we can use the check method, and if returns False, allow to user open the downloader.
- b) Or just remove the message that prompts RHVoice needs to be set as synthesizer, plus I need some testing if these voices are detected even if RHVoice is not set as synthesizer.

Cheers.

zstanecic commented:
Hi, for now, go with the approach a),
the thing is, that naturally, now there is no RHVoice in the list of speech synthesizers!

rmcpantoja commented:
> Hi, for now, go with the approach a), the thing is, that naturally, now there is no RHVoice in the list of speech synthesizers!

Thanks, fixed it. I changed the approach of the synthDriver, so we can now download voices with any synthesizer.

zstanecic commented:
Okay:
First test results:
Installation of the voices works.
Playing of the voice samples works.
Verified andd tested that the RHVoice works when switched to the newly installed voice.
# Thinks to consider fixing or changing:
* removing of voices is not working,
*  we have the following error: 
ERROR - logHandler._threadExceptHook (19:26:01.956) - Thread-5 (on_download_voice) (23596):
Exception in thread Thread-5 (on_download_voice):
Traceback (most recent call last):
  File "threading.pyc", line 1045, in _bootstrap_inner
  File "threading.pyc", line 982, in run
  File "C:\Users\ninos\AppData\Roaming\nvda\addons\RHVoice\globalPlugins\RHVoice\downloader.py", line 157, in on_download_voice
    self.synth = synthDriverHandler.getSynth()
                 ^^^^^^^^^^^^^^^^^^
NameError: name 'synthDriverHandler' is not defined
Why is it so?
We are still placing it as the NVDA add-on.
* when we are listening to the any of the voice samples, we are not able to stop the playback of the sample. It starts repeating, doubling the sample previously played.
*

rmcpantoja commented:
> Okay: First test results: Installation of the voices works. Playing of the voice samples works. Verified andd tested that the RHVoice works when switched to the newly installed voice.
> 
> # Thinks to consider fixing or changing:
> * removing of voices is not working,
> * we have the following error:
>   ERROR - logHandler._threadExceptHook (19:26:01.956) - Thread-5 (on_download_voice) (23596):
>   Exception in thread Thread-5 (on_download_voice):
>   Traceback (most recent call last):
>   File "threading.pyc", line 1045, in _bootstrap_inner
>   File "threading.pyc", line 982, in run
>   File "C:\Users\ninos\AppData\Roaming\nvda\addons\RHVoice\globalPlugins\RHVoice\downloader.py", line 157, in on_download_voice
>   self.synth = synthDriverHandler.getSynth()
>   ^^^^^^^^^^^^^^^^^^
>   NameError: name 'synthDriverHandler' is not defined
>   Why is it so?
>   We are still placing it as the NVDA add-on.
> * when we are listening to the any of the voice samples, we are not able to stop the playback of the sample. It starts repeating, doubling the sample previously played.

Hi @zstanecic, 
Thanks for your feedback!
I will fix the uninstall voice error soon. That was a consequence of the new approach I am using to handle installed voices from the RHVoice's synthDriver.
Regarding playback of samples, it is easy to toggle play/stop when we select a voice that isn't installed yet, because plays the audio through the URL. But, a minor problem is when a voice is already installed, and synthesizes the test message from the installed voice. I could use the `stop` method from the synthDriver, but I don't know how to make a callback when the voice finishes speaking so the play button changes the state to play/stop propperly.

I will try to read the synthDriverHandler to see if exists something similar to I am looking for.

Any suggestions?

Cheers.
